src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java

```diff
@@ -89,50 +103,92 @@ public boolean start() throws Exception {
-                        throw new IOException("Failed to create directory " + f.getParentFile());
-                    }
-                    try {
-                        return new RemoteOutputStream(Files.newOutputStream(f.toPath(), StandardOpenOption.CREATE, StandardOpenOption.APPEND/*, StandardOpenOption.DSYNC*/));
-                    } catch (InvalidPathException e) {
-                        throw new IOException(e);
-                    }
+
+            @Override
+            public RemoteOutputStream invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
+                f = f.getAbsoluteFile();
+                if (!f.getParentFile().exists() && !f.getParentFile().mkdirs()) {
+                    throw new IOException("Failed to create directory " + f.getParentFile());
+                }
+                try {
+                    OutputStream os = Files.newOutputStream(f.toPath(),
+                            StandardOpenOption.CREATE,
+                            append ? StandardOpenOption.APPEND : StandardOpenOption.TRUNCATE_EXISTING,
+                            StandardOpenOption.WRITE);
+
+                    return new RemoteOutputStream(os);
+                } catch (InvalidPathException e) {
+                    throw new IOException(e);
                 }
-            });
+            }
+
+        }
+
+        private RemoteOutputStream getStream() throws IOException, InterruptedException {
+            if (stream == null) {
+                stream = f.act(new TeeFile(append));
+                // From now on, when resumed and the stream gets recreated append new data.
+                append = true;
+            }
+            return stream;
         }
+
+        private void writeObject(ObjectOutputStream oos) throws IOException, InterruptedException {
+            /**
+             * This gets called either to serialize to allow resume or to transfer to a node.
+             * When resuming, the stream gets recreated.
+             * When transferring the stream needs to transfer as well.
+             */
+            oos.defaultWriteObject();
+            if (Channel.current() == null) {
+                // Serialization for resume.
+                oos.writeBoolean(false);
+            } else {
+                // Serialization for transfer.
+                oos.writeBoolean(true);
+                oos.writeObject(getStream());
+            }
+        }
+
+        private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
+            ois.defaultReadObject();
+            if (ois.readBoolean()) {
+                // Transferred, so the stream got transferred as well.
+                stream = (RemoteOutputStream) ois.readObject();
+            }
+        }
+
+        @SuppressWarnings("rawtypes")
+        @Override
+        public OutputStream decorateLogger(Run build, final OutputStream logger) throws IOException, InterruptedException {
```

Comment:
Would be a lot easier to follow the diff if this were moved back up to where it was.